### ISSUE TITLE
ld.so: $LIB dynamic linking tokens expand to lib instead of lib32 on 32bit

### ISSUE THREAD

**Issue by Atemu** [NixOS/nixpkgs#101597]
**Describe the bug**

Dynamic linking tokens can be put in the library path and ld.so is supposed to fill it in according to the executable's arch.

From to the manpage:

```
       $LIB (or equivalently ${LIB})
              This expands to lib or lib64 depending on the architecture
              (e.g., on x86-64, it expands to lib64 and on x86-32, it
              expands to lib).
```

**To Reproduce**
1. Run a 32bit executable with `$LIB` in place of `lib` or `lib32` in the library path
2. Run the executable
3. It loads from `lib`, not `lib32`

This can be observed by looking at the paths in the log of `LD_DEBUG=libs` or infering it from `LD_PRELOAD` error messages:

```
mkdir /tmp/lib{,32}
touch /tmp/lib/64 /tmp/lib32/32
LD_PRELOAD=/tmp/\$LIB/32 ${pkgs.pkgsi386Linux.coreutils}/bin/ls
ERROR: ld.so: object '/tmp/$LIB/32' from LD_PRELOAD cannot be preloaded (cannot open shared object file): ignored.
LD_PRELOAD=/tmp/\$LIB/64 ${pkgs.pkgsi386Linux.coreutils}/bin/ls
ERROR: ld.so: object '/tmp/$LIB/64' from LD_PRELOAD cannot be preloaded (file too short): ignored.
```

(I'm sure there are better methods.)

**Expected behavior**

This would be expected behaviour with 64bit executables but with 32bit ones, it should be the other way around:

```
ERROR: ld.so: object '/tmp/$LIB/32' from LD_PRELOAD cannot be preloaded (file too short): ignored.
ERROR: ld.so: object '/tmp/$LIB/64' from LD_PRELOAD cannot be preloaded (cannot open shared object file): ignored.
```

**Screenshots**

**Additional context**

This popped up in https://github.com/NixOS/nixpkgs/issues/84686 while trying to update to 0.5.1 where Mangohud switched to a loading mechanism which relies on dynamic linking tokens.

**Notify maintainers**

@vcunat @jtojnar @edolstra @Ma27 

**Metadata**
Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

Maintainer information:
```yaml
# a list of nixpkgs attributes affected by the problem
attribute:
# a list of nixos modules affected by the problem
module:
```

**Comment by Atemu** [comment]
Still important to me.

**Comment by MetaDark** [comment]
Yeah, I don't think it would be a good idea to change this globally. To emulate what other distros do, I think we would have to build a separate ld specifically for running 32bit application on a 64bit system that's different from the ld used on a native 32bit system.

Luckily, for vkBasalt & MangoHud, `vulkan-loader` now allows defining multiple layer files with the same name, but different architectures: https://github.com/NixOS/nixpkgs/issues/101597. vkBasalt already assumes this works, but vulkan-loader still needs to be updated in nixpkgs. I'm actually testing this out the upgrade right now, and I'll be submitting a PR shortly.

**Comment by vcunat** [comment]
There is some "multi-lib stdenv" but I don't know the status:
```nix
{
  # A stdenv capable of building 32-bit binaries.  On x86_64-linux,
  # it uses GCC compiled with multilib support; on i686-linux, it's
  # just the plain stdenv.
  stdenv_32bit = lowPrio (if stdenv.hostPlatform.is32bit then stdenv else multiStdenv);
}
```

**Comment by MetaDark** [comment]
@vcunat Hmm, I think the 32bit ld in "multi-lib stdenv" might still expand `$LIB` to lib, but I will try it out.
